scripts: add script to sync kernel changes

anakryiko opened the pull request:
This script automates the process of applying libbpf-relevant changes
from kernel repository on top of current state of libbpf repository.

It uses CHECKPOINT-COMMIT file to keep track of last commit in kernel
repo up to which libbpf is in sync with. If there are any new libbpf
changes in kernel repository, script extracts them, preserving original
commit metadata. It also creates a "sync commit" using cover letter as
a template, which nicely summarizes changes since last sync with kernel.

Usage: `./scripts/sync-kernel.sh <linux-repo> <libbpf-repo>`

If it succeeds, script will create a bunch of local commits in
<libbpf-repo> in separate branch, which can be easily pushed into github
to create a pull request.

Script tries to clean up after itself, except in case of failure. But it
doesn't clean up timestamped branches it creates in both kernel and
libbpf repositories for now. We can add that later.

Signed-off-by: Andrii Nakryiko <andriin@fb.com>

yonghong-song commented on scripts/sync-kernel.sh:
```diff
@@ -0,0 +1,96 @@
+echo "PATCHSET: /tmp/libbpf-${SUFFIX}.patchset"
+
+cd ${LIBBPF_REPO}
+git co -b libbpf-sync-${SUFFIX}
```
I assume this is `git checkout`?

anakryiko replied:
Ah, yes, it's my local alias. I'm so used to it that didn't even realize it's not standard feature, will fix.

yonghong-song commented:
Thanks for implementation. This is in the right direction for automation!

I tried to add a few artificial patches in bpf-next and got the following failures:
```
Ref 'refs/heads/libbpf-2019-02-17T01-17-25.489Z' was rewritten
Rewrite a63440df54bf4159b9389a70450579f990095c29 (3/3) (0 seconds passed, remaining 0 predicted)    
Ref 'refs/heads/libbpf-2019-02-17T01-17-25.489Z' was rewritten
/tmp/libbpf-2019-02-17T01-17-25.489Z.patches/0000-cover-letter.patch
/tmp/libbpf-2019-02-17T01-17-25.489Z.patches/0001-fix1.patch
/tmp/libbpf-2019-02-17T01-17-25.489Z.patches/0002-fix-3.patch
[libbpf-2019-02-17T01-17-25.489Z b5ea6613c83e] sync: latest libbpf changes from kernel
 1 file changed, 1 insertion(+)
 create mode 100644 CHECKPOINT-COMMIT
PATCHSET: /tmp/libbpf-2019-02-17T01-17-25.489Z.patchset
Switched to a new branch 'libbpf-sync-2019-02-17T01-17-25.489Z'
Applying: fix1
error: patch failed: src/btf.h:4
error: src/btf.h: patch does not apply
Patch failed at 0001 fix1
Use 'git am --show-current-patch' to see the failed patch
When you have resolved this problem, run "git am --continue".
If you prefer to skip this patch, run "git am --skip" instead.
To restore the original branch and stop patching, run "git am --abort".
-bash-4.4$
```

The command line I tried:
```
-bash-4.4$ ./scripts/sync-kernel.sh /home/yhs/work/net-next .
```
in the libbpf repo directory.

anakryiko commented:
Can you publish full output and patches you applied on top of bpf-next, I'll try to repro locally. Also what was HEAD of libbpf and kernel?
This happened to me only when master on libbpf and linux repos were out of sync (thus patches didn't apply clean). When I was testing I was even able to re-apply quite a long history of commits on top of some older commits in libbpf, but the trickiest part was correlating commits in linux and libbpf.

yonghong-song commented:
The kernel part:
```
commit 6bcf3db00a5931ccbd78730d85ab6124f826d9ba (HEAD -> copy)
Author: Yonghong Song <yhs@fb.com>
Date:   Sat Feb 16 17:09:37 2019 -0800

    patch3
    
    Signed-off-by: Yonghong Song <yhs@fb.com>

commit add7beebf02a190ccd74ac4a221105090408e9e9
Author: Yonghong Song <yhs@fb.com>
Date:   Sat Feb 16 17:09:03 2019 -0800

    fix 3
    
    Signed-off-by: Yonghong Song <yhs@fb.com>

commit 0ccb09e4c26f3d0189f689b3596f6c4b6655a35a
Author: Yonghong Song <yhs@fb.com>
Date:   Sat Feb 16 17:08:33 2019 -0800

    fix1
    
    Signed-off-by: Yonghong Song <yhs@fb.com>

commit b251f9f63a3bea7864bf627960926d978e97814d (bpf-next/master, bpf-next2)
Author: Peter Oskolkov <posk@google.com>
Date:   Fri Feb 15 09:51:35 2019 -0800

    bpf: make LWTUNNEL_BPF dependent on INET
```
Individual commits:
```
-bash-4.4$ git show 6bcf3db00a5931ccbd78730d85ab6124f826d9ba
commit 6bcf3db00a5931ccbd78730d85ab6124f826d9ba (HEAD -> copy)
Author: Yonghong Song <yhs@fb.com>
Date:   Sat Feb 16 17:09:37 2019 -0800

    patch3
    
    Signed-off-by: Yonghong Song <yhs@fb.com>

diff --git a/t.c b/t.c
new file mode 100644
index 000000000000..f349bb94c37e
--- /dev/null
+++ b/t.c
@@ -0,0 +1 @@
+iinwe filees
-bash-4.4$ git show add7beebf02a190ccd74ac4a221105090408e9e9
commit add7beebf02a190ccd74ac4a221105090408e9e9
Author: Yonghong Song <yhs@fb.com>
Date:   Sat Feb 16 17:09:03 2019 -0800

    fix 3
    
    Signed-off-by: Yonghong Song <yhs@fb.com>

diff --git a/tools/lib/bpf/btf.c b/tools/lib/bpf/btf.c
index ade1c32fb083..e96428c9a47b 100644
--- a/tools/lib/bpf/btf.c
+++ b/tools/lib/bpf/btf.c
@@ -13,6 +13,8 @@
 #include "libbpf.h"
 #include "libbpf_util.h"
 
+#error no
+
 #define max(a, b) ((a) > (b) ? (a) : (b))
 #define min(a, b) ((a) < (b) ? (a) : (b))
 
-bash-4.4$ git show 0ccb09e4c26f3d0189f689b3596f6c4b6655a35a
commit 0ccb09e4c26f3d0189f689b3596f6c4b6655a35a
Author: Yonghong Song <yhs@fb.com>
Date:   Sat Feb 16 17:08:33 2019 -0800

    fix1
    
    Signed-off-by: Yonghong Song <yhs@fb.com>

diff --git a/tools/lib/bpf/btf.h b/tools/lib/bpf/btf.h
index 94bbc249b0f1..c57d42385806 100644
--- a/tools/lib/bpf/btf.h
+++ b/tools/lib/bpf/btf.h
@@ -4,6 +4,8 @@
 #ifndef __LIBBPF_BTF_H
 #define __LIBBPF_BTF_H
 
+#error no
+
 #include <linux/types.h>
 
 #ifdef __cplusplus
-bash-4.4$
```

yonghong-song commented:
Next time I will remember to go to a repo I can push to github.

anakryiko commented:
What was sha on libbpf side?

yonghong-song commented:
libbpf side
```
commit 8521d84455f7bfd17b088ea642767e370939185a (HEAD -> script)
Author: Andrii Nakryiko <andriin@fb.com>
Date:   Sat Feb 16 14:57:53 2019 -0800

    scripts: add script to sync kernel changes
    
    This script automates the process of applying libbpf-relevant changes
    from kernel repository on top of current state of libbpf repository.
    
    It uses CHECKPOINT-COMMIT file to keep track of last commit in kernel
    repo up to which libbpf is in sync with. If there are any new libbpf
    changes in kernel repository, script extracts them, preserving original
    commit metadata. It also creates a "sync commit" using cover letter as
    a template, which nicely summarizes changes since last sync with kernel.
    
    Usage: ./scripts/sync-kernel.sh <linux-repo> <libbpf-repo>
    
    If it succeeds, script will create a bunch of local commits in
    <libbpf-repo> in separate branch, which can be easily pushed into github
    to create a pull request.
    
    Script tries to clean up after itself, except in case of failure. But it
    doesn't clean up timestamped branches it creates in both kernel and
    libbpf repositories for now. We can add that later.
    
    Signed-off-by: Andrii Nakryiko <andriin@fb.com>

commit f0bcba631dec4540fc6ab2cd0a0923a111cf4cf2
Author: yonghong-song <ys114321@gmail.com>
Date:   Thu Feb 7 21:58:07 2019 -0800

    sync with latest bpf-next (#10)
    
    sync with latest bpf-next. tested with fb internal testcase and bcc.
    
    Signed-off-by: Yonghong Song <yhs@fb.com>

commit b19c6dcf623a7adc9e538ddbe2964c2f58dd2417
Author: Yonghong Song <yhs@fb.com>
Date:   Mon Feb 4 19:42:31 2019 -0800

    sync with bpf-next
    
    Sync with the following bug fix:
      commit a8a1f7d09cfc7e18874786c7634c9e71384fcd4e (HEAD -> bpf-next2, bpf-next/master)
      Author: Stanislav Fomichev <sdf@google.com>
      Date:   Mon Feb 4 16:20:55 2019 -0800
    
        libbpf: fix libbpf_print
    
        With the recent print rework we now have the following problem:
        pr_{warning,info,debug} expand to __pr which calls libbpf_print.
        libbpf_print does va_start and calls __libbpf_pr with va_list argument.
        In __base_pr we again do va_start. Because the next argument is a
        va_list, we don't get correct pointer to the argument (and print noting
        in my case, I don't know why it doesn't crash tbh).
    
        ......
    
        Signed-off-by: Stanislav Fomichev <sdf@google.com>
    
    Signed-off-by: Yonghong Song <yhs@fb.com>
```

anakryiko commented:
ah, I get it, it's the logic of updating CHECKPOINT-COMMIT file, it never exists in linux repo, I have a simpler idea on how to handle that. Thanks for catching this!

anakryiko commented:
ahah, so there are two problems:
1. CHECKPOINT-COMMIT should be updated/commited in libbpf repo, not in linux repo (adding it in linux makes it harder to generate correct patch).
2. You used "." as libbpf-repo, script doesn't handle that properly. Try with absolute path for now, I'll think how to properly fix relative paths meanwhile.

I fixed first in updated pull request. Will fix second soon.

BTW, if you are going to try again, keep in mind that you might need to do "git am --abort" in linux repo, and checkout your old HEAD (copy branch).

anakryiko commented:
Ok, fixed second problem as well by switching back to original working dir and then doing `cd LIBBPF_REPO`. Now works as expected:

```
Ref 'refs/heads/libbpf-2019-02-17T02-25-37.047Z' was rewritten
Rewrite 146c3786b73a6b20c524feaebd7196203dea5533 (3/3) (0 seconds passed, remaining 0 predicted)
Ref 'refs/heads/libbpf-2019-02-17T02-25-37.047Z' was rewritten
/tmp/libbpf-2019-02-17T02-25-37.047Z.patches/0000-cover-letter.patch
/tmp/libbpf-2019-02-17T02-25-37.047Z.patches/0001-fix-1.patch
/tmp/libbpf-2019-02-17T02-25-37.047Z.patches/0002-fix-3.patch
PATCHSET: /tmp/libbpf-2019-02-17T02-25-37.047Z.patchset
Switched to a new branch 'libbpf-sync-2019-02-17T02-25-37.047Z'
Applying: fix 1
Applying: fix 3
[libbpf-sync-2019-02-17T02-25-37.047Z 022af37] sync: latest libbpf changes from kernel
 1 file changed, 1 insertion(+), 1 deletion(-)
Success! 2 commits applied.
```

State after running script:

```
commit 022af3793929ce6ab4abd1f178b4b67e03039fba (HEAD -> libbpf-sync-2019-02-17T02-25-37.047Z)
Author: Andrii Nakryiko <andriin@fb.com>
Date:   Sat Feb 16 18:26:12 2019 -0800

    sync: latest libbpf changes from kernel

    Syncing latest libbpf commits from kernel repository.
    Baseline commit:   789f6bab849e04ea029c09b81dc8401dc0268cf9
    Checkpoint commit: 6911e801fa1da0e54b200af8df6908b9fc1c391d

    Andrii Nakryiko (2):
      fix 1
      fix 3

     src/btf.c | 2 ++
     src/btf.h | 2 ++
     2 files changed, 4 insertions(+)

    --
    2.17.1

commit 7825e77a9f04bd272b6aa6e5bf6ffdd239361da5
Author: Andrii Nakryiko <andriin@fb.com>
Date:   Sat Feb 16 17:44:23 2019 -0800

    fix 3

commit 480c282161177b66931fe5175e88115f7394f791
Author: Andrii Nakryiko <andriin@fb.com>
Date:   Sat Feb 16 17:41:06 2019 -0800

    fix 1

commit 88d35a3d80b52017bba32ecb3a4a9d412ac45a9d (kernel-sync-script)
Author: Andrii Nakryiko <andriin@fb.com>
Date:   Sat Feb 16 14:57:53 2019 -0800

    scripts: add script to sync kernel changes

<SNIP>
```

4ast commented on scripts/sync-kernel.sh:
```diff
@@ -0,0 +1,101 @@
+git rebase --onto ${SQUASHED_BASELINE_COMMIT} ${BASELINE_COMMIT} libbpf-${SUFFIX}
+
+# Move all libbpf files into libbpf-projection directory
+git filter-branch --prune-empty -f --tree-filter '                                                                             \
```
didn't know such git command exist :)

anakryiko replied:
it's a "with great power comes great responsibility" type of command :)

anakryiko commented:
@yonghong-song, you can check how sync commit looks like here: https://github.com/libbpf/libbpf/compare/master...anakryiko:libbpf-sync-2019-02-17T02-58-38.109Z?expand=1 

I'm not yet creating that pull request, because we should probably land this one first and then I'll rebase (otherwise we'll have conflicts on CHECKPOINT-COMMIT file).

yonghong-song commented:
I got a failure like below:
```
-bash-4.4$ ./scripts/sync-kernel.sh /home/yhs/work/net-next /home/yhs/work/libbpf
WORKDIR:     /home/yhs/work/libbpf
LINUX REPO:  /home/yhs/work/net-next
LIBBPF REPO: /home/yhs/work/libbpf
SUFFIX: 2019-02-17T06-04-31.687Z
BASELINE COMMIT: 789f6bab849e04ea029c09b81dc8401dc0268cf9
CHECKPOINT COMMIT: 6bcf3db00a5931ccbd78730d85ab6124f826d9ba
SQUASHED BASELINE COMMIT: 81818e2fe001f6e7b23d01ef4e3d9c38655ecfa8
It looks like 'git am' is in progress. Cannot rebase.
-bash-4.4$
```
Any idea what's the issue? I have the same bpf-next hack earlier.

anakryiko commented:
Yeah, due to that initial bug, your linux repository now has `git am` in progress. Just execute `git am --abort && git checkout master` in your linux repo.

yonghong-song commented:
thanks. this time succeeded! Look likes need to learn a few git advanced commands :-)
